.maintain/gitlab/check_dependent_project.sh

```diff
@@ -0,0 +1,262 @@
+    . as $in |
+    paths |
+    select(.[-1]=="source" and . as $p | $in | getpath($p)==null) as $path |
+    del($path[-1]) as $path |
+    $in | getpath($path + ["name"])
+  ')
+  if [ -z "${last_line+_}" ]; then
+    die "No lines were read for cargo metadata of $PWD (some error probably occurred)"
+  fi
+}
+
+match_their_crates() {
+  local target_name="$1"
+  local crates_not_found=()
+  local found
+
+  # workaround for early exits not being detected in command substitution
+  # https://unix.stackexchange.com/questions/541969/nested-command-substitution-does-not-stop-a-script-on-a-failure-even-if-e-and-s
+  local last_line
+
+  # output will be consumed in the format:
+  #   crate
+  #   source
+  #   crate
+  #   ...
+  local next="crate"
+  while IFS= read -r line; do
+    last_line="$line"
+    case "$next" in
+      crate)
+        next="source"
+        crate="$line"
+      ;;
+      source)
+        next="crate"
+        if [ "$line" == "$our_crates_source" ] || [[ "$line" == "$our_crates_source?"* ]]; then
+          for our_crate in "${our_crates[@]}"; do
+            if [ "$our_crate" == "$crate" ]; then
+              found=true
+              break
+            fi
+          done
+          if [ "${found:-}" ]; then
+            unset found
+          else
+            # for avoiding duplicate entries
+            for crate_not_found in "${crates_not_found[@]}"; do
+              if [ "$crate_not_found" == "$crate" ]; then
+                found=true
+                break
+              fi
+            done
+            if [ "${found:-}" ]; then
+              unset found
+            else
+              crates_not_found+=("$crate")
+            fi
+          fi
+        fi
+      ;;
+      *)
+        die "ERROR: Unknown state $next"
+      ;;
+    esac
+  done < <(cargo metadata --quiet --format-version=1 | jq -r '
+    . as $in |
+    paths(select(type=="string")) |
+    select(.[-1]=="source") as $source_path |
+    del($source_path[-1]) as $path |
+    [$in | getpath($path + ["name"]), getpath($path + ["source"])] |
+    .[]
+  ')
+  if [ -z "${last_line+_}" ]; then
+    die "No lines were read for cargo metadata of $PWD (some error probably occurred)"
+  fi
+
+  if [ "${crates_not_found[@]}" ]; then
+    echo -e "Errors during crate matching\n"
+    printf "Failed to detect our crate \"%s\" referenced in $target_name\n" "${crates_not_found[@]}"
+    echo -e "\nNote: this error generally happens if you have deleted or renamed a crate and did not update it in $target_name. Consider opening a companion pull request on $target_name and referencing it in this pull request's description like:\n$target_name companion: [your companion PR here]"
+    die "Check failed"
+  fi
+}
+
+patch_and_check_dependent() {
+  match_their_crates "$(basename "$PWD")"
+  diener patch --crates-to-patch "$this_repo_dir" "$this_repo_diener_arg" --path "Cargo.toml"
+  cargo check --all-targets --workspace
+}
```

Comment:
Not a huge fan of hard-coding the cargo commands to be used universally throughout some repos.
I'd insist on triggering the corresponding repositories' pipelines after getting some outputs from this script and passing them via a trigger job/API.

Reply:
> Not a huge fan of hard-coding the cargo commands to be used universally throughout some repos.

It should work for all the repositories we are targetting (polkadot, cumulus, grandpa-bridge-gadget), thus adding a way of customizing this command would be pointless at this stage because it adds unnecessary complexity for a use-case we do not have.

The only argument for customization I can think of would be if you wanted to trigger a manual pipeline with a custom command, thus it could be like this:

```diff
-cargo check --all-targets --workspace
+${COMPANION_CHECK_COMMAND:-cargo check --all-targets --workspace}
```

It sounds reasonable to me. Would that be an alright implementation?

> I'd insist on triggering the corresponding repositories' pipelines after getting some outputs from this script and passing them via a trigger job/API.

How would that work? The only way I can imagine it, from how this is described, would be using the Gitlab API to trigger a job in the target repository and poll the Gitlab API here until said job finishes. With the current approach we are cloning the companion but, if we were to go with the suggestion, the companion would have to clone Substrate instead (so that it can be patched) and I don't see how that's an improvement at all - it just seems like it would complicate things more.

If it's something along those lines, the implementation would look wildly different from what https://github.com/paritytech/substrate/pull/9010 was doing (and this PR takes after it). Could that suggestion be implemented as a future MR, possibly part of a bigger overhaul? Because I imagine it would be implemented in more repositories other than this and it's a way more drastic change - outside of sending request to the Github API in some places, our scripts are fully local right now, they don't interact with the Gitlab API at all.

Overall I believe dealing with the Gitlab API should be part of a bigger proposal and overhaul. It clearly was not in scope for this PR or https://github.com/paritytech/substrate/pull/9010.

Reply:
The thing is it's going to require more checks on the side of a companion, other jobs, different ones. Since these companions are "modules" for Substrate it's inevitable to grow into triggering projects downstream.

> Could your suggestion be implemented as a future PR, possibly part of a bigger overhaul?

Guess I'll have to. Not everything is prepared for this step so far.